crypto/crypto.c

```diff
@@ -108,6 +110,113 @@ void* crypto_get_signing_cert(void *pkcs7, int cert_num)
 
 }
 
+int crypto_pkcs7_signed_hash_verify(void *pkcs7, void *x509, unsigned char *hash, int hash_len)
+{
+    //currently this function works and the mbedtls version currently perform the following steps
+    //  1. the hash, md context and given x509 are used to generated a signature
+    //  2. all of the signatures in the pkcs7 are compared to the signature generated by the x509
+    //  3. if any of the signatures in the pkcs7 match the genrated signature then return SUCCESS
+    int rc, exp_size, md_nid, num_signers;
+    unsigned char * exp_sig;
+    EVP_PKEY *pk;
+    EVP_PKEY_CTX *pk_ctx;
+    X509_ALGOR *alg;
+    const EVP_MD *evp_md;
+    PKCS7_SIGNER_INFO *signer_info;
+
+    //generate a signature with the x509
+    pk = X509_get_pubkey((X509 *)x509);
+    pk_ctx = EVP_PKEY_CTX_new(pk, NULL);
+    if (pk == NULL || pk_ctx == NULL) {
+        prlog(PR_ERR, "ERROR: Failed to create public key context from x509\n");
+        return CERT_FAIL;       
+    }
+    if (EVP_PKEY_verify_init(pk_ctx) <= 0) {
+        prlog(PR_ERR, "ERROR: Failed to initialize pk context for x509 pk \n");
+        rc = CERT_FAIL;
+        goto out;
+    }
+    if (EVP_PKEY_CTX_set_rsa_padding(pk_ctx, RSA_PKCS1_PADDING) <= 0) {
+        prlog(PR_ERR, "ERROR: Failed to setup pk context with RSA padding\n");
+        rc = CERT_FAIL;
+        goto out;
+    }
+      //extract signer algorithms from pkcs7
+    alg = sk_X509_ALGOR_value(((PKCS7 *)pkcs7)->d.sign->md_algs, 0);
+    if (!alg) {
+        prlog(PR_ERR, "ERROR: Could not extract message digest identifiers from PKCS7\n");
+        rc = PKCS7_FAIL;
+        goto out;        
+    }
+    //extract nid from algorithms
+    md_nid = OBJ_obj2nid(alg->algorithm);
+    //set signature md depending on md in pkcs7
+    switch (md_nid) {
+        case CRYPTO_MD_SHA1:
+            evp_md = EVP_sha1();
+            break;
+        case CRYPTO_MD_SHA224:
+            evp_md = EVP_sha224();
+            break;
+        case CRYPTO_MD_SHA256:
+            evp_md = EVP_sha256();
+            break;
+        case CRYPTO_MD_SHA384:
+            evp_md = EVP_sha384();
+            break;
+        case CRYPTO_MD_SHA512:
+            evp_md = EVP_sha512();
+            break;
+        default:
+            prlog(PR_ERR, "ERROR: Unknown NID (%d) for MD found in PKCS7\n", md_nid);
+            rc = PKCS7_FAIL;
+            goto out;
+    }
```

Comment:
surely there must be an openssl function that can do this for you? there is for mbedtls...

Reply:
EVP_get_digestbynid is used a bit further down the file, can that be used?